codegen/src/main/java/software/amazon/awssdk/codegen/poet/waiters/AsyncWaiterClassSpec.java

```diff
@@ -0,0 +1,122 @@
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.codegen.poet.waiters;
+
+import static javax.lang.model.element.Modifier.FINAL;
+import static javax.lang.model.element.Modifier.PRIVATE;
+import static javax.lang.model.element.Modifier.STATIC;
+
+import com.squareup.javapoet.ClassName;
+import com.squareup.javapoet.FieldSpec;
+import com.squareup.javapoet.MethodSpec;
+import com.squareup.javapoet.ParameterizedTypeName;
+import com.squareup.javapoet.TypeSpec;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import javax.lang.model.element.Modifier;
+import software.amazon.awssdk.codegen.model.intermediate.IntermediateModel;
+import software.amazon.awssdk.codegen.model.intermediate.OperationModel;
+import software.amazon.awssdk.codegen.poet.PoetExtensions;
+import software.amazon.awssdk.core.internal.waiters.WaiterAttribute;
+import software.amazon.awssdk.core.waiters.AsyncWaiter;
+import software.amazon.awssdk.core.waiters.WaiterResponse;
+import software.amazon.awssdk.utils.ThreadFactoryBuilder;
+
+public class AsyncWaiterClassSpec extends BaseWaiterClassSpec {
+
+    private final PoetExtensions poetExtensions;
+    private final ClassName className;
+    private final IntermediateModel model;
+    private final String modelPackage;
+
+    public AsyncWaiterClassSpec(IntermediateModel model) {
+        super(model, ClassName.get(AsyncWaiter.class));
+        this.model = model;
+        this.modelPackage = model.getMetadata().getFullModelPackageName();
+        this.poetExtensions = new PoetExtensions(model);
+        this.className = poetExtensions.getAsyncWaiterClass();
+    }
+
+    @Override
+    public ClassName className() {
+        return className;
+    }
+
+    @Override
+    protected ClassName clientClassName() {
+        return poetExtensions.getClientClass(model.getMetadata().getAsyncInterface());
+    }
+
+    @Override
+    protected ParameterizedTypeName getWaiterResponseType(OperationModel opModel) {
+        ClassName pojoResponse = ClassName.get(modelPackage, opModel.getReturnType().getReturnType());
+        ParameterizedTypeName waiterResponse = ParameterizedTypeName.get(ClassName.get(WaiterResponse.class), pojoResponse);
+
+        return ParameterizedTypeName.get(ClassName.get(CompletableFuture.class),
+                                         waiterResponse);
+    }
+
+    @Override
+    protected ClassName interfaceClassName() {
+        return poetExtensions.getAsyncWaiterInterface();
+    }
+
+    @Override
+    protected Optional<String> additionalWaiterConfig() {
+        return Optional.of(".scheduledExecutorService(executorService)");
+    }
+
+    @Override
+    protected void additionalConstructorInitialization(MethodSpec.Builder method) {
+        method.beginControlFlow("if (builder.executorService == null)")
```

Comment:
Hmm why does the async waiter need a threadpool? Would it be enough to just rely on the async client?

Reply:
We still need a `ScheduledExecutorService` to schedule the attempts.

Reply:
Ah gotcha forgot about the backoffs